src/evaluators/response/PHAL_ResponseSquaredL2DifferenceSide_Def.hpp

```diff
@@ -200,46 +196,52 @@ evaluateFields(typename Traits::EvalData workset)
             sum += sq * w_measure(sideSet_idx,qp);
           }
 
-          this->local_response_eval(cell, 0) = sum*scaling;
-          this->global_response_eval(0) += sum*scaling;
-        }
+          this->local_response_eval(cell,0) = sum*scaling;
+          KU::atomic_add<ExecutionSpace>(&(this->global_response_eval(0)), sum*scaling);
+        });
         break;
       case 1:
-        for (int sideSet_idx = 0; sideSet_idx < sideSet.size; ++sideSet_idx)
-        {
+        Kokkos::parallel_for(this->getName(),RangePolicy(0,sideSet.size),
+                       KOKKOS_CLASS_LAMBDA(const int& sideSet_idx) {
           // Get the local data of cell
-          const int cell = sideSet.ws_elem_idx.h_view(sideSet_idx);
+          const int cell = sideSet.ws_elem_idx.d_view(sideSet_idx);
+
+          ScalarT diff_1[8] = {0};
 
           ScalarT sum = 0;
           for (int qp=0; qp<numQPs; ++qp)
           {
             ScalarT sq = 0;
             RealType rms2 = rmsScaling ? std::pow(rootMeanSquareField(sideSet_idx,qp),2) : 1.0;
-            // Computing squared difference at qp
-            // Precompute differentce and access fields only n times (not n^2)
-            for (size_t i=0; i<dims[2]; ++i)
+
+            for (size_t i=0; i<dims_2; ++i)
               diff_1[i] = sourceField(sideSet_idx,qp,i) - (target_value ? target_value_val : targetField(sideSet_idx,qp,i));
 
             if(isFieldGradient){
-              for (int i=0; i<sideDim; ++i)
-                for (int j=0; j<sideDim; ++j)
+              for (int i=0; i<sideDim; ++i) {
+                for (int j=0; j<sideDim; ++j) {
                   sq += diff_1[i]*metric(sideSet_idx,qp,i,j)*diff_1[j];
+                }
+              }
             } else {
-              for (size_t i=0; i<dims[2]; ++i)
+              for (size_t i=0; i<dims_2; ++i) {
                 sq += diff_1[i]*diff_1[i];
+              }
             }
             sum += sq / rms2 * w_measure(sideSet_idx,qp);
           }
 
-          this->local_response_eval(cell, 0) = sum*scaling;
-          this->global_response_eval(0) += sum*scaling;
-        }
+          this->local_response_eval(cell,0) = sum*scaling;
+          KU::atomic_add<ExecutionSpace>(&(this->global_response_eval(0)), sum*scaling);
```

Comment:
Wouldn't it be better to use parallel_reduce, rather than a parallel_for with atomic access? All threads access _the same_ value, so contention is very high here...

Reply:
Yeah, this should definitely be a reduce. I ported this early on when I was just trying to get something to work. I'll fix it.

Reply:
@bartgol Is there a trick to doing a parallel_reduce with Sacado FAD types? I created a very simple example program based on https://kokkos.org/kokkos-core-wiki/ProgrammingGuide/Custom-Reductions-Built-In-Reducers-with-Custom-Scalar-Types.html and I'm getting compiler errors:

```
/pscratch/sd/m/mcarlson/IntroToHPC/repos/trilinos-lite/packages/sacado/example/my_custom_reduce_example.cpp(35): error: no instance of constructor "Kokkos::View<DataType, Properties...>::View [with DataType=ScalarT *, Properties=<Kokkos::CudaSpace, Kokkos::MemoryUnmanaged>]" matches the argument list
            argument types are: (ScalarT *, int)
          detected during:
            instantiation of "SumScalarT<ScalarT, Space>::result_view_type SumScalarT<ScalarT, Space>::view() const [with ScalarT=ScalarT, Space=Kokkos::CudaSpace]"
/pscratch/sd/m/mcarlson/IntroToHPC/repos/trilinos-lite/packages/kokkos/core/src/Kokkos_Parallel_Reduce.hpp(1525): here
            instantiation of "void Kokkos::Impl::ParallelReduceAdaptor<PolicyType, FunctorType, ReturnType>::execute_impl(const std::string &, const PolicyType &, const FunctorType &, ReturnType &) [with PolicyType=Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>, FunctorType=lambda [](int, ValueType &)->void, ReturnType=SumScalarT<ScalarT, Kokkos::CudaSpace>]"
/pscratch/sd/m/mcarlson/IntroToHPC/repos/trilinos-lite/packages/kokkos/core/src/Kokkos_Parallel_Reduce.hpp(1542): here
            instantiation of "std::enable_if_t<<expression>, void> Kokkos::Impl::ParallelReduceAdaptor<PolicyType, FunctorType, ReturnType>::execute(const std::string &, const PolicyType &, const FunctorType &, ReturnType &) [with PolicyType=Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>, FunctorType=lambda [](int, ValueType &)->void, ReturnType=SumScalarT<ScalarT, Kokkos::CudaSpace>, Dummy=SumScalarT<ScalarT, Kokkos::CudaSpace>]"
/pscratch/sd/m/mcarlson/IntroToHPC/repos/trilinos-lite/packages/kokkos/core/src/Kokkos_Parallel_Reduce.hpp(1798): here
            instantiation of "std::enable_if_t<<expression>, void> Kokkos::parallel_reduce(const size_t &, const FunctorType &, const ReturnType &) [with FunctorType=lambda [](int, ValueType &)->void, ReturnType=SumScalarT<ScalarT, Kokkos::CudaSpace>]"
(51): here
```

Reply:
Uhm, I am not sure. Maybe @etphipp has some advice.